mountpoint-s3/src/store.rs

```diff
@@ -0,0 +1,262 @@
+use std::fmt::Debug;
+
+use async_trait::async_trait;
+use futures::{executor::ThreadPool, task::Spawn};
+use mountpoint_s3_client::{
+    error::{
+        DeleteObjectError, GetObjectAttributesError, GetObjectError, HeadObjectError, ListObjectsError,
+        ObjectClientError, PutObjectError,
+    },
+    types::{
+        DeleteObjectResult, ETag, GetObjectAttributesResult, HeadObjectResult, ListObjectsResult, ObjectAttribute,
+        ObjectClientResult, PutObjectParams,
+    },
+    ObjectClient, PutObjectRequest,
+};
+use thiserror::Error;
+
+use crate::{
+    checksums::{ChecksummedBytes, IntegrityError},
+    data_cache::DataCache,
+    prefetch::{
+        cached_feed::{CacheKey, CachedPartFeed},
+        feed::ClientPartFeed,
+        PrefetcherConfig,
+    },
+};
+use crate::{
+    prefetch::{self, feed::ObjectPartFeed, Prefetcher},
+    sync::Arc,
+};
+
+/// A generic interface to S3-like object store.
+/// Similar to [ObjectClient], but provides a [ObjectStore::prefetch] method instead
+/// of [ObjectClient::get_object].
+#[async_trait]
+pub trait ObjectStore: Clone {
```

Comment:
Would it make more sense to do `pub trait ObjectStore: ObjectClient` rather than duplicating all the methods?

Reply:
The downside I see is that if `ObjectStore` extends `ObjectClient`, it will expose `get_object` as well: fine for the default implementation, but what about the "caching" implementation? Will it have to implement caching for it? Or just accept that `get_object` and `prefetch` may behave differently?